### PR TITLE
ref #314

### PR DESCRIPTION
**Pull request by ThomasRetornaz**
Add missing saturated sadd ssub
* int8,uint8,int16,uint16,int32,uint32,int64,uint64,float,double
* sse2/sse4
* avx/avx2
* avx512
* fallback
* neon
Test:
* Validated on unbuntu 1804, g++ 7,4 and Visual 2019
* Critical to review AVX ssub unint32_t uint64_t not work, i can't figure out why
* Other Problems not critical
** AVX512 flavor doesnt compil under visual 2019
** Mandatory add_definitions(-D_SILENCE_TR1_NAMESPACE_DEPRECATION_WARNING) for compiling  gtest under visual
** Mandatory BUILD_SHARED_LIBS=ON on windows
** flaky error with sse_ under visual
** abort/corruption in visual debug (maybe a problem with SECURE_*)

### CODE REVIEW

File: include/xsimd/types/xsimd_avx512_int32.hpp [hunk 1]
```diff
@@ -349,6 +388,43 @@ namespace xsimd
...
+                *
+                * MIT License see https://github.com/agenium-scale/nsimd/blob/master/LICENSE
+                * ====================================================
+                */
+                //todo bench againt unrroled loop
+                //todo factorize int32_t uint32_t
+                const auto ures = _mm512_add_epi32(lhs, rhs);
+                const auto umax = _mm512_set1_epi32(std::numeric_limits<uint32_t>::max());
+                const auto is_overflow = _mm512_cmp_epu32_mask(_mm512_add_epi32(lhs, umax), _mm512_add_epi32(ures, umax), _MM_CMPINT_NLE);
+                return _mm512_mask_blend_epi32(is_overflow, umax, ures);
```
Comment (JohanMabille) [inline include/xsimd/types/xsimd_avx512_int32.hpp]: I think that this should be `_mm512_mask_blend_epi32(is_overflow, ures, umax);`, according to [this](https://software.intel.com/sites/landingpage/IntrinsicsGuide/#expand=821,282,4952,721,435&text=_mm512_mask_blend_epi32). I agree that this is totally counter-intuitive.

File: include/xsimd/types/xsimd_avx512_int32.hpp [hunk 2]
```diff
@@ -349,6 +388,43 @@ namespace xsimd
...
+                * MIT License see https://github.com/agenium-scale/nsimd/blob/master/LICENSE
+                * ====================================================
+                */
+                //todo bench againt unrroled loop
+                //todo factorize int32_t uint32_t
+                const auto ures = _mm512_sub_epi32(lhs, rhs);
+                const auto cte = _mm512_set1_epi32(std::numeric_limits<uint32_t>::max());
+                const auto is_underflow = _mm512_cmp_epu32_mask(_mm512_add_epi32(rhs, cte), _mm512_add_epi32(lhs, cte), _MM_CMPINT_NLE);
+                const auto umin = _mm512_set1_epi32(std::numeric_limits<uint32_t>::lowest());
+                return _mm512_mask_blend_epi32(is_underflow,umin, ures);
```
Comment (JohanMabille) [inline include/xsimd/types/xsimd_avx512_int32.hpp]: Same fremark here.

**Comment by ThomasRetornaz** [comment]
My bad for AVX512, wrong copy paste ... We could also note that i don't simplify the case of avx512.uint64_t. I will double check with intel sde
Do i push my commit as is or do i have to rebase or amend last commit?
Thanks a lot for the review
Regards
TR

**Comment by JohanMabille** [comment]
As you wish. You can squash (or amend now) or we can do it later when the PR is ready to be merged.

File: include/xsimd/types/xsimd_avx_int32.hpp [hunk 3]
```diff
@@ -368,6 +413,50 @@ namespace xsimd
...
+                //todo factorize int32_t uint32_t
+
+
+                const auto ures = _mm256_add_epi32(lhs, rhs);
+                const auto umax = _mm256_set1_epi32(std::numeric_limits<uint32_t>::max());
+                const auto is_overflow = _mm256_cmpgt_epi32(_mm256_add_epi32(lhs, umax), _mm256_add_epi32(ures, umax));
+                return _mm256_blendv_epi8(ures, umax, is_overflow);
+#else
+
+                XSIMD_APPLY_SSE_FUNCTION(sse_int32_batch_kernel<uint32_t>::sadd, lhs, rhs);
```
Comment (ThomasRetornaz) [inline include/xsimd/types/xsimd_avx_int32.hpp]: This code fail don't know why

File: include/xsimd/types/xsimd_avx_int32.hpp [hunk 4]
```diff
@@ -368,6 +413,50 @@ namespace xsimd
...
+
+
+                const auto ures = _mm256_sub_epi32(lhs, rhs);
+                const auto cte = _mm256_set1_epi32(std::numeric_limits<uint32_t>::max());
+                const auto is_underflow = _mm256_cmpgt_epi32(_mm256_add_epi32(rhs, cte), _mm256_add_epi32(lhs, cte));
+                const auto umin = _mm256_set1_epi32(std::numeric_limits<uint32_t>::lowest());
+                return _mm256_blendv_epi8(ures, umin, is_underflow);
+#else
+
+                 XSIMD_APPLY_SSE_FUNCTION(sse_int32_batch_kernel<uint32_t>::ssub, lhs, rhs);
```
Comment (ThomasRetornaz) [inline include/xsimd/types/xsimd_avx_int32.hpp]: idem

File: include/xsimd/types/xsimd_avx_int64.hpp [hunk 5]
```diff
@@ -186,6 +187,48 @@ namespace xsimd
...
+                //todo bench againt unrroled loop
+                //todo factorize int_64/uint_64
+
+                const auto ures = _mm256_add_epi64(lhs, rhs);
+                const auto umax = _mm256_set1_epi64x(std::numeric_limits<uint64_t>::max());
+                const auto is_overflow = _mm256_cmpgt_epi64(_mm256_add_epi64(lhs, umax), _mm256_add_epi64(ures, umax));
+                return _mm256_blendv_epi8(ures, umax, is_overflow);
+#else
+
+                XSIMD_APPLY_SSE_FUNCTION(sse_int64_batch_kernel<uint64_t>::sadd, lhs, rhs);
```
Comment (ThomasRetornaz) [inline include/xsimd/types/xsimd_avx_int64.hpp]: idem

File: include/xsimd/types/xsimd_avx_int64.hpp [hunk 6]
```diff
@@ -186,6 +187,48 @@ namespace xsimd
...
+
+
+                const auto ures = _mm256_sub_epi64(lhs, rhs);
+                const auto cte = _mm256_set1_epi64x(std::numeric_limits<uint64_t>::max());
+                const auto is_underflow= _mm256_cmpgt_epi64(_mm256_add_epi64(rhs, cte),_mm256_add_epi64(lhs, cte));
+                const auto umin = _mm256_set1_epi64x(std::numeric_limits<uint64_t>::lowest());
+                return _mm256_blendv_epi8( ures,umin,is_underflow);
+#else
+
+                XSIMD_APPLY_SSE_FUNCTION(sse_int64_batch_kernel<uint64_t>::ssub, lhs, rhs);
```
Comment (ThomasRetornaz) [inline include/xsimd/types/xsimd_avx_int64.hpp]: idem

**Comment by serge-sans-paille** [comment]
@ThomasRetornaz there hasn't been a lot of activity on that PR recently, is it okay if I update it, I'll (obviously) respect proper attribution of your original work.

**Comment by ThomasRetornaz** [comment]
@serge-sans-paille yes of course 🙂
Yes if i remind correctly all work except two avx corner cases
I can t figure out why 🙁
Thanks
Regards
TR

**Comment by JohanMabille** [comment]
Superseded by #454 .

Thanks again @ThomasRetornaz  for the original implementation!

**Pull request closed without merging**